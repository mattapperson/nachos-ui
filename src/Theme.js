import React, { Component } from "react";
import PropTypes from "prop-types";
import { StyleSheet } from "react-native";
import defaultBranding from "./branding_defaults";

export const ThemeContext = React.createContext({
  theme: {}
});

export class Provider extends React.Component {
  state = {};
  static getDerivedStateFromProps(nextProps, prevState) {
    if (JSON.stringify(nextProps.theme) !== JSON.stringify(prevState)) {
      //return this.createTheme(nextProps.theme);
    }
    return null;
  }

  themeConfigs = {};
  branding = defaultBranding;

  updateThemeConfig = (name, themeConfig) => {
    themeConfig = JSON.parse(
      JSON.stringify(themeConfig).replace(/@([\w_-]+)/gm, (match, key) => {
        if (!this.branding[key]) console.log(key);

        return this.branding[key];
      })
    );

    // if (name === "H1") {
    //   console.log(themeConfig);
    // }
    return (this.themeConfigs[name] = {
      computedStyle: themeConfig.style
        ? StyleSheet.create(themeConfig.style)
        : {},
      props: themeConfig.settings
    });
  };

  render() {
    return (
      <ThemeContext.Provider
        value={{
          theme: this.themeConfigs,
          updateThemeConfig: this.updateThemeConfig
        }}
      >
        {this.props.children}
      </ThemeContext.Provider>
    );
  }
}

export function withTheme(componentName, ThemedComponent) {
  return class extends React.Component {
    constructor() {
      super();
      try {
        ThemeContext._currentValue.updateThemeConfig(
          componentName,
          ThemedComponent.themeConfig
        );
      } catch (e) {
        console.warn(`failed to update theme for ${componentName}`, e);
      }
    }

    isObject(item) {
      return item && typeof item === "object" && !Array.isArray(item);
    }

    mergeDeep(target, ...sources) {
      if (!sources.length) return target;
      const source = sources.shift();

      if (this.isObject(target) && this.isObject(source)) {
        for (const key in source) {
          if (this.isObject(source[key])) {
            if (!target[key]) Object.assign(target, { [key]: {} });
            this.mergeDeep(target[key], source[key]);
          } else {
            Object.assign(target, { [key]: source[key] });
          }
        }
      }

      return this.mergeDeep(target, ...sources);
    }
    render() {
      return (
        <ThemeContext.Consumer>
          {({ theme }) => {
            const computedTheme =
              (theme[componentName] || {}).computedStyle || {};

            const fullTheme = this.props.theme
              ? this.mergeDeep({}, computedTheme, this.props.theme)
              : computedTheme;

            return (
              <ThemedComponent
                {...(theme[componentName] || {}).props || {}}
                {...this.props}
                theme={fullTheme}
              />
            );
          }}
        </ThemeContext.Consumer>
      );
    }
  };
}
