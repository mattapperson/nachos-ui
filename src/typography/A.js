import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { Linking } from 'react-native'
import Text from './Text'
import themeManager from '../themeManager'

const defaultTheme = {
  ...themeManager.getStyle('Text'),
  FONT_COLOR: '#2f8cff',
  PADDING_VERTICAL: 0,
}

themeManager.setSource('A', () => defaultTheme)

const defaultStyle = (theme) => {
  return {
    color: theme.FONT_COLOR,
    textDecorationLine: 'underline',
    textDecorationStyle: 'solid',
    textDecorationColor: theme.FONT_COLOR,
  }
}

class A extends Component {
  static propTypes = {
    href: PropTypes.string,
    onPress: PropTypes.func,
    onError: PropTypes.func,
    children: PropTypes.node,
    style: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
    theme: PropTypes.object,
  }

  static defaultProps = {
    onPress: () => {},
    onError: () => {},
  }

  _handlePress = (e) => {
    if (this.props.onPress) {
      this.props.onPress(e)
      if (e.defaultPrevented) {
        return
      }
    }

    if (!this.props.href) {
      return
    }

    Linking.canOpenURL(this.props.href).then((supported) => {
      if (supported) {
        Linking.openURL(this.props.href)
      } else {
        this.props.onError(
          `Don't know how to open URI: ${this.props.href}`
        )
      }
    })
  }

  render() {
    const theme = this.props.theme || themeManager.getStyle('A')
    const baseStyle = defaultStyle(theme)
    return (
      <Text
        accessibilityRole='link'
        style={[baseStyle, this.props.style]}
        onPress={this._handlePress}
      >
        {this.props.children}
      </Text>
    )
  }
}

export default A
