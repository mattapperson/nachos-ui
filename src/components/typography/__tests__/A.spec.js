import "react-native";
import React from "react";
import renderer from "react-test-renderer";

import A from "../A";
import { Provider } from "../../Theme";

it("should render correctly", () => {
  const component = renderer.create(
    <Provider>
      <A href="http://avocode.com">This is a link</A>
    </Provider>
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
