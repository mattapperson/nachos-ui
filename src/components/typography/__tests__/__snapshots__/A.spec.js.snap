// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<a
  className="rn-backgroundColor-1niwhzg rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-cursor-1loqt21 rn-display-1471scf rn-fontFamily-baq3cm rn-fontSize-1b43r93 rn-fontStyle-13uqrnb rn-fontVariant-ebii48 rn-height-1pexk7n rn-lineHeight-rjixqe rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-textAlign-fdjqy7 rn-whiteSpace-q42fyq rn-wordWrap-qvutc0"
  data-focusable={true}
  dir="auto"
  onClick={[Function]}
  onKeyDown={[Function]}
  role="link"
  style={
    Object {
      "color": "rgba(121,69,239,1.00)",
      "fontWeight": "normal",
      "textDecoration": "underline solid rgba(121,69,239,1.00)",
    }
  }
>
  This is a link
</a>
`;
